packages/transaction-controller/src/utils/history.ts

```diff
@@ -40,16 +56,99 @@ export function updateTransactionHistory(
 
   const currentState = snapshotFromTransactionMeta(transactionMeta);
   const previousState = replayHistory(transactionMeta.history);
-  const historyEntry = generateHistoryEntry(previousState, currentState, note);
+  const newHistoryEntry = generateHistoryEntry(
+    previousState,
+    currentState,
+    note,
+  );
+
+  if (newHistoryEntry.length > 0) {
+    // Casts required here because this list has two separate types of entries:
+    // TransactionMeta and TransactionHistoryEntry. The only TransactionMeta is the first
+    // entry, but TypeScript loses that type information when `slice` is called for some reason.
+    let updatedHistory = transactionMeta.history.slice() as TransactionHistory;
+    updatedHistory.push(newHistoryEntry);
+
+    if (updatedHistory.length > MAX_HISTORY_LENGTH) {
+      updatedHistory = compressTransactionHistory(updatedHistory);
+    }
 
-  if (historyEntry.length > 0) {
     return merge({}, transactionMeta, {
-      history: [...transactionMeta.history, historyEntry],
+      history: updatedHistory,
     });
   }
   return transactionMeta;
 }
 
+/**
+ * Compress the transaction history, if it is possible to do so without compressing entries used
+ * for display. History entries are merged together to make room for a single new entry.
+ *
+ * @param transactionHistory - The transaction history to compress.
+ * @returns A compressed transaction history.
+ */
+function compressTransactionHistory(
+  transactionHistory: TransactionHistory,
+): TransactionHistory {
+  const initialEntry = transactionHistory[0];
+  // Casts required here because this list has two separate types of entries:
+  // TransactionMeta and TransactionHistoryEntry. The only TransactionMeta is the first
+  // entry, but TypeScript loses that type information when `slice` is called for some reason.
+  const historyEntries = transactionHistory.slice(
+    1,
+  ) as TransactionHistoryEntry[];
+
+  const firstNonDisplayedEntryIndex = historyEntries.findIndex(
```

Comment:
Forgive me if I'm misunderstanding, but we're looking for the first entry without any of the activity paths, and merging it (ideally) with the next entry.

But in a worst case scenario, could this have negligible impact on size if both entries update entirely different fields with different operations?

If we assume we likely reach the history limit when some kind of loop starts adding duplicate entries, then would it be more impactful to start merging from the end? Since that's more likely to reduce the net state size?

Reply:
>But in a worst case scenario, could this have negligible impact on size if both entries update entirely different fields with different operations?

True, but there are only so many fields (certainly fewer than 100, which is the limit set at the moment). They will have to overlap eventually, at which point it'll save space.

100 or even 1000 entries won't have a huge impact on memory usage, these are not large entries. This is really just meant to stop growth. It's certainly not optimal, optimal would be to stop using this for display, hopefully we'll do that soon.

>If we assume we likely reach the history limit when some kind of loop starts adding duplicate entries, then would it be more impactful to start merging from the end? Since that's more likely to reduce the net state size?

In that scenario, starting at the beginning still gets us there eventually. I figured that starting at the beginning would leave more useful diagnostic information (e.g. the more recent actions would be more relevant to what the user did when they took a state snapshot). But I'm not really sure.